sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -84,19 +84,50 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
     }
   }
 
+  // Extract a list of logical plans to be joined for join-order comparisons.
+  // Since `ExtractFiltersAndInnerJoins` handles left-deep trees only, this function have
+  // the same strategy to extract the plan list.
+  private[optimizer] def extractLeftDeepInnerJoins(plan: LogicalPlan)
+    : Seq[LogicalPlan] = plan match {
+    case j @ Join(left, right, _: InnerLike, _) => right +: extractLeftDeepInnerJoins(left)
+    case p @ Project(_, j @ Join(_, _, _: InnerLike, _)) => extractLeftDeepInnerJoins(j)
+    case _ => Seq(plan)
+  }
+
+  private def sameJoinOrder(plan1: LogicalPlan, plan2: LogicalPlan): Boolean = {
+    extractLeftDeepInnerJoins(plan1) == extractLeftDeepInnerJoins(plan2)
+  }
+
+  private def mayCreateOrderedJoin(
+      originalPlan: LogicalPlan,
+      input: Seq[(LogicalPlan, InnerLike)],
+      conditions: Seq[Expression]): LogicalPlan = {
+    val orderedJoins = createOrderedJoin(input, conditions)
+    if (!sameJoinOrder(orderedJoins, originalPlan)) {
+      if (originalPlan.output != orderedJoins.output) {
```

Comment:
I think this also fixes an existing bug for join reordering, right? Shall we add it to the issue description?

Reply:
What's the bug you pointed out here?

Reply:
I mean the original join re-order before your fix did not apply a Project and that must be a bug, as is illustrated by your change to JoinReorderSuite. I think we should file it and mark it fixed by this PR.

Reply:
ah, yes, I got you. btw, we need to file a separate jira? Is it bad to add the bug in the description of this jira and this pr?

Reply:
I think opening another JIRA would be better, for people discovering this bug in earlier versions are more likely to find it if it is filed as an independent issue.